esp/bindings/http/platform/httpservice.cpp

```diff
@@ -974,3 +875,538 @@ int CEspHttpServer::onGet()
+{
+    VStringBuffer sessionIDStr("%u", sessionID);
+    addCookie(authBinding->querySessionIDCookieName(), sessionIDStr.str(), authBinding->getSessionTimeoutSeconds());
+    addCookie(SESSION_AUTH_COOKIE, "1", 0); //time out when browser is closed
+    m_response->redirect(*m_request, authBinding->queryLoginURL());
+}
+
+void CEspHttpServer::send200OK()
+{
+    m_response->setContentType("text/html; charset=UTF-8");
+    m_response->setStatus(HTTP_STATUS_OK);
+    m_response->send();
+}
+
+unsigned CEspHttpServer::createHTTPSession(EspAuthRequest& authReq, const char* loginURL)
+{
+    Owned<IRemoteConnection> conn = getSDSConnection(authReq.authBinding->queryDomainSessionSDSPath(), RTM_LOCK_WRITE, SESSION_SDS_LOCK_TIMEOUT);
+    return createHTTPSession(authReq, conn->queryRoot(), loginURL);
+}
+
+unsigned CEspHttpServer::createHTTPSession(EspAuthRequest& authReq, IPropertyTree* domainSessions, const char* _loginURL)
+{
+    CDateTime now;
+    now.setNow();
+    time_t createTime = now.getSimple();
+
+    StringBuffer peer;
+    VStringBuffer idStr("%s_%ld", m_request->getPeer(peer).str(), createTime);
+    unsigned sessionID = hashc((unsigned char *)idStr.str(), idStr.length(), 0);
+
+    VStringBuffer xpath("%s[%s='%u']", PathSessionSession, PropSessionID, sessionID);
+    IPropertyTree* sessionTree = domainSessions->queryBranch(xpath.str());
+    if (sessionTree)
+    {
+        sessionTree->setPropInt64(PropSessionLastAccessed, createTime);
+        return sessionID;
+    }
+    ESPLOG(LogMax, "New sessionID <%d> at <%ld> in createHTTPSession()", sessionID, createTime);
+
+    Owned<IPropertyTree> ptree = domainSessions->addPropTree(PathSessionSession);
+    ptree->setProp(PropSessionNetworkAddress, peer.str());
+    ptree->setPropInt64(PropSessionID, sessionID);
+    ptree->setPropInt(PropSessionState, HTTPSS_new);
+    ptree->setPropInt64(PropSessionCreateTime, createTime);
+    ptree->setPropInt64(PropSessionLastAccessed, createTime);
+
+    if (!isEmptyString(_loginURL))
+        ptree->setProp(PropSessionLoginURL, _loginURL);
+    else
+    {
+        StringBuffer loginURL = authReq.httpPath;
+        if (authReq.requestParams && authReq.requestParams->hasProp("__querystring"))
+            loginURL.append("?").append(authReq.requestParams->queryProp("__querystring"));
+        if (!loginURL.isEmpty() && streq(loginURL.str(), "/WsSMC/"))
+            ptree->setProp(PropSessionLoginURL, "/");
+        else
+            ptree->setProp(PropSessionLoginURL, loginURL.str());
+    }
+    return sessionID;
+}
+
+void CEspHttpServer::removeTimedOutDomainSessions(EspHttpBinding* authBinding, IPropertyTree* domainSessions)
+{
+    //Removing HTTPSessions if timed out
+    CDateTime now;
+    now.setNow();
+    time_t timeNow = now.getSimple();
+
+    ICopyArrayOf<IPropertyTree> toRemove;
+    Owned<IPropertyTreeIterator> iter = domainSessions->getElements(PathSessionSession);
+    ForEach(*iter)
```

Comment:
@jakesmith @wangkx  does this iterate over every session in this domain every time we authorize based on a session id?  Would it be better to do this intermittently for cleanup, but check our own timeout every time?